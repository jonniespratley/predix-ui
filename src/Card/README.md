The `<Card/>` component consist of a header area and content area. The header can optionally contain a descriptive icon in the top left corner and/or an actions area in the top right corner for settings or other action icons. Cards are incredibly flexible in terms of what content can be placed in the main area of the card and how it's arranged.


## Usage

```react
const { Card } = PxReact;
<div>
  <Card headerText='My Card'>
    This is the main content area of a Predix card.
  </Card>
</div>
```

### With loading

```react
state:
  loading: true
---
const { Card, Spinner } = PxReact;
const LoadingCard = (props) => {
  const {loading, children} = props;
  return (
    <Card headerText='Loading Card' loading>
      {!loading && children}
      {loading && <div className='flex flex--center'><Spinner size={75}/></div>}
    </Card>
  );
};

<LoadingCard loading={state.loading}>
  <p>This card should stop loading after a few seconds</p>
</LoadingCard>

```

### With custom CSS

```react
const { Card } = PxReact;
<div className='custom'>
  <style>{`
    .custom{
      --px-card-background-color: #fff;
      --px-card-border-color: #222;
      --px-card-header-color: #222;
    }
  `}</style>
  <Card headerText='My Card' icon='px-fea:analysis'>
    This is the main content area of a Predix card.
  </Card>
</div>
```

## Properties

```table
span: 6
rows:
  - Name: headerText
    Type: string
    Description: Text to be displayed in the header of the card.
  - Name: icon
    Type: string
    Description: Icon that appears to the left of the card header text. Should be a valid icon name from the Predix icon set.
```

### Styling
The following custom properties are available for styling:

```table
span: 6
rows:
  - Custom Property: --px-card-background-color
    Description: Background color of the card

  - Custom Property: --px-card-border-color
    Description: Border color for the top of the card

  - Custom Property: --px-card-header-color
    Description: Color of the header text

  - Custom Property: --px-card-icon-color
    Description: Color of the card's main icon'

  - Custom Property: --px-card-action-icon-color
    Description: Color of icons in actions slot

  - Custom Property: --px-card-action-icon-color--hover
    Description: Color of icons in actions slot when hovered

  - Custom Property: --px-card-action-icon-color--pressed
    Description: Color of icons in actions slot when pressed
```
