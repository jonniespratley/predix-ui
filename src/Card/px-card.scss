@import 'px-defaults-design/_settings.defaults.scss'; // Component

.px-card {
  display         : block;
  width           : 100%;
  background-color: var(--px-card-background-color, lightgray);
  padding-bottom  : 4 * $inuit-base-spacing-unit;
  border-top      : 1px solid var(--px-card-border-color, gray);

  &[full-bleed] .contents {
    padding: 0;
  }

  header {
    padding      : $inuit-base-spacing-unit--large $inuit-base-spacing-unit $inuit-base-spacing-unit;
    border-bottom: 1px solid var(--px-card-border-color, transparent);
    margin-bottom: $inuit-base-spacing-unit;

    .epsilon {
      height     : 2rem;
      line-height: 2rem;
      color      : var(--px-card-header-color, black);
    }

    i {
      color : var(--px-card-action-icon-color, black);
      cursor: pointer;

      &:hover {
        color: var(--px-card-action-icon-color--hover, blue);
      }

      &:active {
        color: var(--px-card-action-icon-color--pressed, blue);
      }

      &:not(:last-of-type) {
        margin-right: 0.66666rem;
      }
    }
  }

  #headericon {
    height: 2rem;
    width : 2rem;
    color : var(--px-card-icon-color, black);
    cursor: auto;
  }

  section {
    padding-left : $inuit-base-spacing-unit;
    padding-right: $inuit-base-spacing-unit;
  }
}