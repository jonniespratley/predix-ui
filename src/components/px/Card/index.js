import React from 'react';
import stylesheet from './px-card.scss';
import classnames from 'classnames';
import Icon from '../IconSet/px-icon';

const baseClasses = classnames('px-card');

export default ({
  headerText,
  icon,
  style,
  children
}) => (
  <div className={baseClasses} style={style}>
    {headerText && <header className='flex'>
      {icon && <span className='icon'><Icon icon={icon} size={22}/></span>}
      <span className='epsilon caps'>{headerText}</span>
    </header>}
    <section>{children}</section>
    <style jsx>{stylesheet}</style>
  </div>
);
