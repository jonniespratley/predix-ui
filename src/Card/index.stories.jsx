import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean } from '@storybook/addon-knobs/react';
import { withReadme } from 'storybook-readme';

import README from './README.md';
import Card from '.';

const customActions = () => (
  <div>
    <button type="button">Edit</button>
  </div>
);

storiesOf('Components / Card', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(README))
  .addWithJSX('default', () => (
    <Card
      headerText={text('headerText', 'My Card')}
      fullBleed={boolean('fullBleed', false)}
    >
      {text('content', 'This is the main context area of the card.')}
    </Card>
  ))
  .addWithJSX('with icon', () => (
    <Card
      icon={text('icon', 'px-fea:analysis')}
      headerText={text('headerText', 'My Card')}
      fullBleed={boolean('fullBleed', false)}
    >
      {text('content', 'This is the main context area of the card.')}
    </Card>
  ))
  .addWithJSX('with actions', () => (
    <Card
      icon={text('icon', 'px-fea:analysis')}
      headerText={text('headerText', 'My Card')}
      fullBleed={boolean('fullBleed', false)}
      actions={customActions}
    >
      {text('content', 'This is the main context area of the card.')}
    </Card>
  ));
