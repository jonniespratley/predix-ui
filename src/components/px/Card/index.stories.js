import React from 'react';
import { storiesOf } from '@storybook/react';
import Card from './';

storiesOf('Card', module)
	.add('default', () => (
		<Card/>
	));
