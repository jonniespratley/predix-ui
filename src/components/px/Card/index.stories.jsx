import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs/react';
import { withReadme } from 'storybook-readme';

import README from './README.md';
import Card from './';

const customActions = () => (
  <div>Expand</div>
);

storiesOf('Components / Card', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(README))
  .addWithJSX('default', () => (
    <Card
      headerText={text('headerText', 'My Card')}
    >
      {text('content', 'This is the main context area of the card.')}
    </Card>
  ))
  .addWithJSX('with icon', () => (
    <Card
      icon={text('icon', 'px-fea:analysis')}
      headerText={text('headerText', 'My Card')}
    >
      {text('content', 'This is the main context area of the card.')}
    </Card>
  ))
  .addWithJSX('with actions', () => (
    <Card
      icon={text('icon', 'px-fea:analysis')}
      headerText={text('headerText', 'My Card')}
      actions={customActions}
    >
      {text('content', 'This is the main context area of the card.')}
    </Card>
  ));

