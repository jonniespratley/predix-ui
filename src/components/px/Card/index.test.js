import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Card from './';

describe('Card', () => {
  test('should render', () =>{
    const tree = renderer.create(<Card headerText='Card Header' icon='px-nav:home'>Content</Card>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render with headerText', () =>{
    const wrapper = shallow(<Card headerText='Card Title'/>);
    expect(wrapper.find('.px-card-header')).toHaveLength(1);
  });

  test('should render with icon', () =>{
    const wrapper = shallow(<Card headerText='Card Title' icon='px-nav:home'/>);
    expect(wrapper.find('.px-card-icon')).toHaveLength(1);
  });

  test('should render with children', () =>{
    const wrapper = shallow(<Card headerText='Card Title' icon='px-nav:home'>This is content</Card>);
    expect(wrapper.find('.px-card-body')).toHaveLength(1);
  });
});
