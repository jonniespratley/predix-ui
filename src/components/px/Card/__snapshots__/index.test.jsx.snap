// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render 1`] = `
<div
  className="px-card hLbjWt"
>
  <header
    className="px-card-header frRFri"
  >
    <span
      className="yYkWD"
    >
      <i
        className="px-card-icon fxynje"
      >
        <i
          className="px-icon px-nav-home femNjm"
          color={null}
          style={
            Object {
              "height": 22,
              "width": 22,
            }
          }
        >
          <svg
            dangerouslySetInnerHTML={
              Object {
                "__html": "<g id=\\"px-nav-home\\"><path stroke-linejoin=\\"round\\" d=\\"M3.5 11v10.5h15V11m-18-.5L11 1l10.5 9.5\\"/><path stroke-linejoin=\\"round\\" d=\\"M8.5 15.5h5v6h-5z\\"/></g>",
              }
            }
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "display": "block",
                "height": 22,
                "pointerEvents": "none",
                "width": 22,
              }
            }
            viewBox="0 0 22 22"
          />
        </i>
      </i>
      Card Header
    </span>
  </header>
  <section
    className="px-card-body keWRp"
  >
    Content
  </section>
</div>
`;
